File: aegis/backend/test/internal/utils/sanitization_test.go
package utils_test

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"

	"github.com/balkanid/aegis-backend/internal/utils"
)

func TestSanitizeHTML(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "no special chars",
			input:    "hello world",
			expected: "hello world",
		},
		{
			name:     "ampersand",
			input:    "Tom & Jerry",
			expected: "Tom & Jerry",
		},
		{
			name:     "less than",
			input:    "5 < 10",
			expected: "5 < 10",
		},
		{
			name:     "greater than",
			input:    "10 > 5",
			expected: "10 > 5",
		},
		{
			name:     "quotes",
			input:    `"Hello"`,
			expected: `"Hello"`,
		},
		{
			name:     "apostrophe",
			input:    "Don't",
			expected: "Don&#x27;t",
		},
		{
			name:     "forward slash",
			input:    "path/to/file",
			expected: "path&#x2F;to&#x2F;file",
		},
		{
			name:     "backtick",
			input:    "`code`",
			expected: "&#x60;code&#x60;",
		},
		{
			name:     "equals",
			input:    "a=b",
			expected: "a&#x3D;b",
		},
		{
			name:     "mixed special chars",
			input:    `<script>alert("XSS")</script>`,
			expected: `<script>alert(&#x22;XSS&#x22;)</script>`,
		},
		{
			name:     "unicode characters",
			input:    "café résumé naïve",
			expected: "café résumé naïve",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := utils.SanitizeHTML(tt.input)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestSanitizeFilename(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "valid filename",
			input:    "document.pdf",
			expected: "document.pdf",
		},
		{
			name:     "dangerous chars",
			input:    `file<>:|?"*\.txt`,
			expected: "file.txt",
		},
		{
			name:     "control chars",
			input:    "file\x00\x01\x1f.txt",
			expected: "file.txt",
		},
		{
			name:     "leading dots",
			input:    "...file.txt",
			expected: "file.txt",
		},
		{
			name:     "trailing dots",
			input:    "file.txt...",
			expected: "file.txt",
		},
		{
			name:     "multiple spaces",
			input:    "file   with    spaces.txt",
			expected: "file with spaces.txt",
		},
		{
			name:     "long filename",
			input:    strings.Repeat("a", 300) + ".txt",
			expected: strings.Repeat("a", 255),
		},
		{
			name:     "only dangerous chars",
			input:    `<>:|?"*\`,
			expected: "",
		},
		{
			name:     "spaces and dots only",
			input:    "   ...   ",
			expected: "",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := utils.SanitizeFilename(tt.input)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestSanitizeUserInput(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "normal text",
			input:    "Hello World",
			expected: "Hello World",
		},
		{
			name:     "with tabs and newlines",
			input:    "Hello\tWorld\nTest",
			expected: "Hello\tWorld\nTest",
		},
		{
			name:     "control chars removed",
			input:    "Hello\x00\x01\x1fWorld",
			expected: "HelloWorld",
		},
		{
			name:     "leading/trailing spaces",
			input:    "  Hello World  ",
			expected: "Hello World",
		},
		{
			name:     "long input truncated",
			input:    strings.Repeat("a", 1200),
			expected: strings.Repeat("a", 1000),
		},
		{
			name:     "carriage return preserved",
			input:    "Line1\r\nLine2",
			expected: "Line1\r\nLine2",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := utils.SanitizeUserInput(tt.input)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestSanitizeSearchQuery(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "normal query",
			input:    "search term",
			expected: "search term",
		},
		{
			name:     "html chars removed",
			input:    `<script>alert("xss")</script>`,
			expected: "scriptalertxss/script",
		},
		{
			name:     "quotes removed",
			input:    `"quoted" 'single'`,
			expected: "quoted single",
		},
		{
			name:     "ampersand removed",
			input:    "Tom & Jerry",
			expected: "Tom  Jerry",
		},
		{
			name:     "whitespace trimmed",
			input:    "  search term  ",
			expected: "search term",
		},
		{
			name:     "long query truncated",
			input:    strings.Repeat("a", 150),
			expected: strings.Repeat("a", 100),
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := utils.SanitizeSearchQuery(tt.input)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestIsValidEmail(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected bool
	}{
		{"empty string", "", false},
		{"valid email", "test@example.com", true},
		{"valid email with subdomain", "test@sub.example.com", true},
		{"valid email with numbers", "test123@example.com", true},
		{"valid email with dots", "test.email@example.com", true},
		{"valid email with plus", "test+tag@example.com", true},
		{"invalid no at", "testexample.com", false},
		{"invalid no domain", "test@", false},
		{"invalid no local", "@example.com", false},
		{"invalid spaces", "test @ example.com", false},
		{"invalid special chars", "test<>@example.com", false},
		{"invalid double at", "test@@example.com", false},
		{"invalid no tld", "test@example", false},
		{"invalid single char tld", "test@example.c", false},
		{"too long email", strings.Repeat("a", 250) + "@example.com", false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := utils.IsValidEmail(tt.input)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestIsValidUsername(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected bool
	}{
		{"empty string", "", false},
		{"too short", "ab", false},
		{"valid username", "testuser", true},
		{"valid with underscore", "test_user", true},
		{"valid with hyphen", "test-user", true},
		{"valid with numbers", "user123", true},
		{"valid mixed", "user_123-test", true},
		{"too long", strings.Repeat("a", 51), false},
		{"invalid spaces", "test user", false},
		{"invalid special chars", "test@user", false},
		{"invalid dots", "test.user", false},
		{"invalid uppercase", "TestUser", false},
		{"starts with underscore", "_testuser", true},
		{"starts with hyphen", "-testuser", true},
		{"starts with number", "1testuser", true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := utils.IsValidUsername(tt.input)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestSanitizeSQLLikeQuery(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "no special chars",
			input:    "search term",
			expected: "search term",
		},
		{
			name:     "percent escaped",
			input:    "50%",
			expected: "50\\%",
		},
		{
			name:     "underscore escaped",
			input:    "test_case",
			expected: "test\\_case",
		},
		{
			name:     "both chars",
			input:    "test%_case",
			expected: "test\\%\\_case",
		},
		{
			name:     "multiple occurrences",
			input:    "%test%_case%",
			expected: "\\%test\\%\\_case\\%",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := utils.SanitizeSQLLikeQuery(tt.input)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestNormalizeWhitespace(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "no whitespace",
			input:    "hello",
			expected: "hello",
		},
		{
			name:     "single spaces",
			input:    "hello world",
			expected: "hello world",
		},
		{
			name:     "multiple spaces",
			input:    "hello   world",
			expected: "hello world",
		},
		{
			name:     "tabs",
			input:    "hello\tworld",
			expected: "hello world",
		},
		{
			name:     "newlines",
			input:    "hello\nworld",
			expected: "hello world",
		},
		{
			name:     "mixed whitespace",
			input:    "hello \t\n  world",
			expected: "hello world",
		},
		{
			name:     "leading/trailing whitespace",
			input:    "  hello world  ",
			expected: "hello world",
		},
		{
			name:     "only whitespace",
			input:    " \t\n ",
			expected: "",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := utils.NormalizeWhitespace(tt.input)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestRemoveNonPrintable(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "printable chars",
			input:    "Hello World 123!@#",
			expected: "Hello World 123!@#",
		},
		{
			name:     "control chars removed",
			input:    "Hello\x00\x01\x1fWorld",
			expected: "HelloWorld",
		},
		{
			name:     "unicode printable",
			input:    "café résumé naïve",
			expected: "café résumé naïve",
		},
		{
			name:     "mixed printable and non-printable",
			input:    "Hello\x00World\x1fTest",
			expected: "HelloWorldTest",
		},
		{
			name:     "only non-printable",
			input:    "\x00\x01\x1f",
			expected: "",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := utils.RemoveNonPrintable(tt.input)
			assert.Equal(t, tt.expected, result)
		})
	}
}

File: aegis/frontend/src/test/utils/sanitization.test.ts
import {
  sanitizeHtml,
  sanitizeFilename,
  sanitizeUserInput,
  isValidEmail,
  isValidUsername,
  sanitizeSearchQuery,
  isValidFileSize,
  isValidMimeType,
} from '../../utils/sanitization';

describe('Sanitization Utils', () => {
  describe('sanitizeHtml', () => {
    it('should encode HTML entities to prevent XSS', () => {
      const input = '<script>alert("xss")</script><img src="x" onerror="alert(1)">';
      const expected = '<script>alert("xss")<&#x2F;script><img src&#x3D;"x" onerror&#x3D;"alert(1)">';

      expect(sanitizeHtml(input)).toBe(expected);
    });

    it('should handle all dangerous characters', () => {
      const input = '&<>"\'/`=';
      const expected = '&<>"&#x27;&#x2F;&#x60;&#x3D;';

      expect(sanitizeHtml(input)).toBe(expected);
    });

    it('should return empty string for non-string input', () => {
      expect(sanitizeHtml(null as any)).toBe('');
      expect(sanitizeHtml(undefined as any)).toBe('');
      expect(sanitizeHtml(123 as any)).toBe('');
    });

    it('should handle empty strings', () => {
      expect(sanitizeHtml('')).toBe('');
    });

    it('should preserve safe content', () => {
      const input = 'Hello World! This is safe content.';
      expect(sanitizeHtml(input)).toBe(input);
    });
  });

  describe('sanitizeFilename', () => {
    it('should remove dangerous characters', () => {
      const input = 'file<>:"/\\|?*name.txt';
      const expected = 'filename.txt';

      expect(sanitizeFilename(input)).toBe(expected);
    });

    it('should remove leading and trailing dots', () => {
      expect(sanitizeFilename('...file.txt')).toBe('file.txt');
      expect(sanitizeFilename('file.txt...')).toBe('file.txt');
    });

    it('should normalize whitespace', () => {
      expect(sanitizeFilename('file   with    spaces.txt')).toBe('file with spaces.txt');
    });

    it('should trim whitespace', () => {
      expect(sanitizeFilename('  file.txt  ')).toBe('file.txt');
    });

    it('should limit filename length', () => {
      const longName = 'a'.repeat(300) + '.txt';
      const result = sanitizeFilename(longName);
      expect(result.length).toBeLessThanOrEqual(255);
      expect(result).toBe('a'.repeat(255)); // Should be truncated and trailing dot removed
    });

    it('should return empty string for non-string input', () => {
      expect(sanitizeFilename(null as any)).toBe('');
      expect(sanitizeFilename(undefined as any)).toBe('');
    });

    it('should handle empty strings', () => {
      expect(sanitizeFilename('')).toBe('');
    });

    it('should preserve safe filenames', () => {
      const safeName = 'my_document_123.pdf';
      expect(sanitizeFilename(safeName)).toBe(safeName);
    });
  });

  describe('sanitizeUserInput', () => {
    it('should trim whitespace', () => {
      expect(sanitizeUserInput('  input  ')).toBe('input');
    });

    it('should remove control characters', () => {
      const input = 'input\x00\x01\x02\x03text';
      expect(sanitizeUserInput(input)).toBe('inputtext');
    });

    it('should limit input length', () => {
      const longInput = 'a'.repeat(1200);
      const result = sanitizeUserInput(longInput);
      expect(result.length).toBeLessThanOrEqual(1000);
    });

    it('should return empty string for non-string input', () => {
      expect(sanitizeUserInput(null as any)).toBe('');
      expect(sanitizeUserInput(undefined as any)).toBe('');
    });

    it('should handle empty strings', () => {
      expect(sanitizeUserInput('')).toBe('');
    });
  });

  describe('isValidEmail', () => {
    it('should validate correct email formats', () => {
      expect(isValidEmail('user@example.com')).toBe(true);
      expect(isValidEmail('test.email+tag@domain.co.uk')).toBe(true);
      expect(isValidEmail('user_name@subdomain.domain.org')).toBe(true);
    });

    it('should reject invalid email formats', () => {
      expect(isValidEmail('invalid')).toBe(false);
      expect(isValidEmail('@example.com')).toBe(false);
      expect(isValidEmail('user@')).toBe(false);
      expect(isValidEmail('user@@example.com')).toBe(false);
      expect(isValidEmail('user example.com')).toBe(false);
    });

    it('should reject emails that are too long', () => {
      const longEmail = 'a'.repeat(250) + '@example.com';
      expect(isValidEmail(longEmail)).toBe(false);
    });

    it('should return false for non-string input', () => {
      expect(isValidEmail(null as any)).toBe(false);
      expect(isValidEmail(undefined as any)).toBe(false);
      expect(isValidEmail(123 as any)).toBe(false);
    });
  });

  describe('isValidUsername', () => {
    it('should validate correct usernames', () => {
      expect(isValidUsername('user123')).toBe(true);
      expect(isValidUsername('test_user')).toBe(false); // underscores not allowed
      expect(isValidUsername('UserName')).toBe(true);
    });

    it('should reject usernames that are too short', () => {
      expect(isValidUsername('ab')).toBe(false);
    });

    it('should reject usernames that are too long', () => {
      const longUsername = 'a'.repeat(60);
      expect(isValidUsername(longUsername)).toBe(false);
    });

    it('should reject usernames with special characters', () => {
      expect(isValidUsername('user@name')).toBe(false);
      expect(isValidUsername('user-name')).toBe(false);
      expect(isValidUsername('user.name')).toBe(false);
    });

    it('should return false for non-string input', () => {
      expect(isValidUsername(null as any)).toBe(false);
      expect(isValidUsername(undefined as any)).toBe(false);
    });
  });

  describe('sanitizeSearchQuery', () => {
    it('should remove HTML characters', () => {
      const input = '<script>alert("xss")</script>search query';
      const expected = 'scriptalert(xss)/scriptsearch query';

      expect(sanitizeSearchQuery(input)).toBe(expected);
    });

    it('should trim whitespace', () => {
      expect(sanitizeSearchQuery('  query  ')).toBe('query');
    });

    it('should limit query length', () => {
      const longQuery = 'a'.repeat(150);
      const result = sanitizeSearchQuery(longQuery);
      expect(result.length).toBeLessThanOrEqual(100);
    });

    it('should return empty string for non-string input', () => {
      expect(sanitizeSearchQuery(null as any)).toBe('');
      expect(sanitizeSearchQuery(undefined as any)).toBe('');
    });
  });

  describe('isValidFileSize', () => {
    it('should validate file sizes within limits', () => {
      expect(isValidFileSize(1024)).toBe(true);
      expect(isValidFileSize(1024 * 1024)).toBe(true); // 1MB
      expect(isValidFileSize(100 * 1024 * 1024)).toBe(true); // 100MB (default max)
    });

    it('should reject invalid file sizes', () => {
      expect(isValidFileSize(0)).toBe(false);
      expect(isValidFileSize(-1024)).toBe(false);
      expect(isValidFileSize(101 * 1024 * 1024)).toBe(false); // Over default max
    });

    it('should use custom max size', () => {
      expect(isValidFileSize(2048, 1024)).toBe(false); // Over custom max
      expect(isValidFileSize(512, 1024)).toBe(true); // Under custom max
    });

    it('should reject non-number input', () => {
      expect(isValidFileSize('1024' as any)).toBe(false);
      expect(isValidFileSize(null as any)).toBe(false);
      expect(isValidFileSize(undefined as any)).toBe(false);
    });
  });

  describe('isValidMimeType', () => {
    it('should validate default allowed MIME types', () => {
      expect(isValidMimeType('image/jpeg')).toBe(true);
      expect(isValidMimeType('video/mp4')).toBe(true);
      expect(isValidMimeType('audio/mpeg')).toBe(true);
      expect(isValidMimeType('text/plain')).toBe(true);
      expect(isValidMimeType('application/pdf')).toBe(true);
    });

    it('should reject disallowed MIME types', () => {
      expect(isValidMimeType('application/x-executable')).toBe(false);
      expect(isValidMimeType('application/javascript')).toBe(false);
    });

    it('should validate against custom allowed types', () => {
      const allowedTypes = ['custom/type1', 'custom/type2'];
      expect(isValidMimeType('custom/type1', allowedTypes)).toBe(true);
      expect(isValidMimeType('custom/type2', allowedTypes)).toBe(true);
      expect(isValidMimeType('other/type', allowedTypes)).toBe(false);
    });

    it('should return false for non-string input', () => {
      expect(isValidMimeType(null as any)).toBe(false);
      expect(isValidMimeType(undefined as any)).toBe(false);
      expect(isValidMimeType(123 as any)).toBe(false);
    });

    it('should handle empty allowed types array', () => {
      expect(isValidMimeType('image/jpeg', [])).toBe(true); // Uses default types when empty
      expect(isValidMimeType('application/x-executable', [])).toBe(false);
    });
  });
});